astropy/nddata/nddata.py

```diff
@@ -263,7 +263,8 @@ def __str__(self):
 
     def __repr__(self):
```

Comment:
This is sort of un-related, but the `__repr__` using numpy functions to print the *values* in the array, interacts really badly with non-numpy arrays like Dask.

I would rather see all this code only run if the data and/or masks are only subo;asses of numpy arrays.

Reply:
@Cadair good idea. Somewhat miraculously, if I do: 
```python
import numpy as np
import dask.array as da
from astropy.nddata import NDData

NDData(da.array(np.arange(5)), mask=[False, True, False, True, False])
```
I get 
```
NDData([0 — 2 — 4])
```
Is that ok for you?

Reply:
I don't fully understand what's happening underneath – so it's entirely possible the numpy calls are converting the dask array to a numpy array, and this single line would melt my MacBook if that were a typical DKIST dask array.

Reply:
I fear your are right that as it is, you might run into trouble... But with `Masked`, it should be fairly easy to change, as it overrides `array2string` & co. For large arrays, only the edges are shown, so the conversion to array would have to be moved after those have been selected.

p.s. If you are interested, the override is around https://github.com/astropy/astropy/blob/2aef9dc327fbc9fd29fb2ced2be77eb7f3f625ec/astropy/utils/masked/function_helpers.py#L975

Reply:
I think it definitely will load the dask array, and last I checked it computed the whole dask array, not just the bits it ended up displaying.

I would suggest we include `.data.__repr__` in the `.NDData.__repr__` because then it's agnostic to the array type?

Reply:
I'm happy with that, but it means that the concise masking representation is out of scope. 

Numpy gives us one hint on how to dodge this problem in their masked array repr:
```python
>>> ma
masked_array(
  data=[[--, 2, 3],
        [2, 3, 4],
        [3, --, 5]],
  mask=[[ True, False, False],
        [False, False, False],
        [False,  True, False]],
  fill_value=999999)
```
We could just show each NDData attribute separately. 

Pros: 
* we can rely on each attribute's repr and remain agnostic about `type(nddata.data)`
* it's pretty clear what's in the object
* it's easy to implement

Cons: 
* it's less pretty than either @mhvk's or numpy's implementation, which both obscure masked elements.

Here's an implementation,
```python
prefix = nddata.__class__.__name__ + "("
contents = []
for attr in ['data', 'mask', 'uncertainty', 'unit']:
    attr_data = getattr(nddata, attr)
    if attr_data is not None:
        attr_prefix = f'\n  {attr}='
        attr_repr = repr(attr_data)
        attr_repr = attr_repr.replace("\n", f'\n{" " * (len(attr_prefix) - 1)}')
        contents.append(attr_prefix + attr_repr)
return prefix + ','.join(contents) + ")"
```
and using this implementation we can make an NDData from dask
```python
import numpy as np
import dask.array as da
from astropy.nddata import NDData

nddata = NDData(
    da.array(np.arange(5_000_000).reshape((250000, 20))), 
    mask=np.random.randint(0, 2, size=5_000_000).astype(bool).reshape((250000, 20))
)
```
which prints like this:
```
NDData(
  data=dask.array<array, shape=(250000, 20), dtype=int64, chunksize=(250000, 20), chunktype=numpy.ndarray>,
  mask=array([[ True,  True, False, ..., False, False, False],
              [False, False, False, ...,  True, False,  True],
              [ True,  True, False, ..., False,  True,  True],
              ...,
              [False,  True, False, ...,  True, False, False],
              [False, False, False, ...,  True, False, False],
              [False,  True, False, ..., False,  True, False]]))
```

Thoughts, @Cadair, @mhvk?

Reply:
My sense would be to use `Masked` if `.data` is an `ndarray` subclass and otherwise just `.data.__repr__` as you suggest. Hopefully, we can adjust `Masked` to work on `dask` arrays itself - unlikely `Quantity`, it is meant to be somewhat agnostic to what is being masked.

But I can also see the case for keeping it simple and just rely on the `__repr__` of each part.

Indeed, for a `__repr__` it is generally good to have format that is close to what is needed to initialize. I guess perhaps the bigger question is whether one wants a `__str__` as well, where one would (I think) to not show the mask explicitly (and where `Masked(nd.data << nd.unit, mask=nd.mask)` would do OK...)

Reply:
So, is this resolved or something still needs to be done here?

Reply:
Is this resolved?